v2/pkg/input/uncover.go

```diff
@@ -0,0 +1,152 @@
+	var censysRateLimiter, fofaRateLimiter, shodanRateLimiter, shodanIdbRateLimiter, quakeRatelimiter, hunterRatelimiter, zoomeyeRatelimiter *ratelimit.Limiter
+	if uncoverOptions.Delay > 0 {
+		censysRateLimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		fofaRateLimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		shodanRateLimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		shodanIdbRateLimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		quakeRatelimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		hunterRatelimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+		zoomeyeRatelimiter = ratelimit.New(context.Background(), 1, time.Duration(uncoverOptions.Delay))
+	} else {
+		censysRateLimiter = ratelimit.NewUnlimited(context.Background())
+		fofaRateLimiter = ratelimit.NewUnlimited(context.Background())
+		shodanRateLimiter = ratelimit.NewUnlimited(context.Background())
+		shodanIdbRateLimiter = ratelimit.NewUnlimited(context.Background())
+		quakeRatelimiter = ratelimit.NewUnlimited(context.Background())
+		hunterRatelimiter = ratelimit.NewUnlimited(context.Background())
+		zoomeyeRatelimiter = ratelimit.NewUnlimited(context.Background())
+	}
+	var agents []uncover.Agent
+	// declare clients
+	for _, engine := range uncoverOptions.Engine {
+		var (
+			agent uncover.Agent
+			err   error
+		)
+		switch engine {
+		case "shodan":
+			agent, err = shodan.NewWithOptions(&uncover.AgentOptions{RateLimiter: shodanRateLimiter})
+		case "censys":
+			agent, err = censys.NewWithOptions(&uncover.AgentOptions{RateLimiter: censysRateLimiter})
+		case "fofa":
+			agent, err = fofa.NewWithOptions(&uncover.AgentOptions{RateLimiter: fofaRateLimiter})
+		case "shodan-idb":
+			agent, err = shodanidb.NewWithOptions(&uncover.AgentOptions{RateLimiter: shodanIdbRateLimiter})
+		case "quake":
+			agent, err = quake.NewWithOptions(&uncover.AgentOptions{RateLimiter: quakeRatelimiter})
+		case "hunter":
+			agent, err = hunter.NewWithOptions(&uncover.AgentOptions{RateLimiter: hunterRatelimiter})
+		case "zoomeye":
+			agent, err = zoomeye.NewWithOptions(&uncover.AgentOptions{RateLimiter: zoomeyeRatelimiter})
+		default:
+			err = errors.New("unknown agent type")
+		}
+		if err != nil {
+			return nil, err
+		}
+		agents = append(agents, agent)
+	}
+	// enumerate
+	var wg sync.WaitGroup
+	ret := make(chan string)
+	go func() {
+		for _, q := range uncoverOptions.Query {
+			uncoverQuery := &uncover.Query{
+				Query: q,
+				Limit: uncoverOptions.Limit,
+			}
+			for _, agent := range agents {
+				wg.Add(1)
+				go func(agent uncover.Agent, uncoverQuery *uncover.Query) {
+					defer wg.Done()
+					keys := uncoverOptions.Provider.GetKeys()
+					if keys.Empty() && agent.Name() != "shodan-idb" {
+						gologger.Error().Label(agent.Name()).Msgf("empty keys\n")
```

Comment:
We can give more details in the error and the trailing newline is redundant

Reply:
Updated it to add more details.